### PR TITLE
Add book example that uses publics in asm

### PR DESCRIPTION
**Pull request by leonardoalt**
Fixes https://github.com/powdr-labs/powdr/issues/1225

### CODE REVIEW

File: test_data/asm/sqrt_with_public.asm [hunk 1]
```diff
@@ -0,0 +1,24 @@
+machine Square with degree: 8 {
+    reg pc[@pc];
+    reg X[<=];
+    reg Y[<=];
+    reg A;
+
+    public N = A(7);
+
+    instr j l: label {
```
[the remark this answers is not in the document]
Comment (georgwiese) [reply test_data/asm/sqrt_with_public.asm:9]: But you could make it a loop instruction: `instr loop {pc' = pc}`! I just tried and it works! I guess we check whether it's already there?

[the remark this answers is not in the document]
Comment (leonardoalt) [reply test_data/asm/sqrt_with_public.asm:9]: already where?

File: book/src/publics.md [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+Let's run all steps needed to generate and verify a proof that 3^2 = 9:
+
+1. Setup step:
+```console
+powdr setup 8 --backend halo2 --field bn254
+```
+
+2. Verification Key generation:
+```console
+powdr verification-key test_data/asm/sqrt.asm --field bn254 --backend halo2 --params params.bin
```
Comment (georgwiese) [inline book/src/publics.md:29]: This failed for me. I think the constants need to exist already? worked if I run the `powdr pil` command first.

File: book/src/publics.md [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+```
+
+4. Proof generation:
+```console
+powdr prove test_data/asm/sqrt.asm --field bn254 --backend halo2 --params params.bin --vkey vkey.bin
+```
+
+5. Proof verification:
+```console
+powdr verify test_data/asm/sqrt.asm --field bn254 --backend halo2 --params params.bin --vkey vkey.bin --proof sqrt_proof.bin --publics 9
```
Comment (georgwiese) [inline book/src/publics.md:44]:
```suggestion
powdr verify test_data/asm/sqrt_with_public.asm --field bn254 --backend halo2 --params params.bin --vkey vkey.bin --proof sqrt_with_public_proof.bin --publics 9
```

Also in all other lines!

**Review by georgwiese — COMMENTED** [review]
I ran the tutorial and a few things didn't work for me.

This script now runs through (assumes to be run from the output directory):
```bash
set -e

powdr setup 8 --backend halo2 --field bn254
powdr pil ../test_data/asm/sqrt_with_public.asm --field bn254 -i 3
powdr verification-key ../test_data/asm/sqrt_with_public.asm --field bn254 --backend halo2 --params params.bin
powdr prove ../test_data/asm/sqrt_with_public.asm --field bn254 --backend halo2 --params params.bin --vkey vkey.bin
powdr verify ../test_data/asm/sqrt_with_public.asm --field bn254 --backend halo2 --params params.bin --vkey vkey.bin --proof sqrt_with_public_proof.bin --publics 9
```

How hard would it be to test the command lines in the docs? Could we just include a script file that runs all commands (with comments in between instead of markdown), and also test it in the CI?

File: test_data/asm/sqrt_with_public.asm (hunk 1 above)
[the remark this answers is not in the document]
Comment (georgwiese) [reply test_data/asm/sqrt_with_public.asm:9]: In the process of converting ASM to PIL, powdr adds a `loop` instruction. After a `return`, the "operation ID" is unconstrained, but there is a witgen hint that sets it a line that runs `loop`.

[the remark this answers is not in the document]
Comment (georgwiese) [reply test_data/asm/sqrt_with_public.asm:9]: Oh, I just double-checked. The instruction that powdr adds is called `_loop`, so if you add an instruction called `loop`, you'd have the same instruction twice, under a different name. But that's fine I think and a bit more concise than having a `jump` instruction IMO.

**Comment by georgwiese** [comment]
Oh, wait, I just tried this, and it also works:
```rust
machine Square with degree: 8 {
    reg pc[@pc];
    reg X[<=];
    reg Y[<=];
    reg A;

    public N = A(7);

    instr square X -> Y {
        Y = X * X
    }
	
    function main {
        A <=X= ${ std::prover::Query::Input(0) };
        A <== square(A);
    }
}
```

So seems like we can just leave out the `return` and then execution will directly continue with the infinite loop, without resetting the registers, which is exactly what we want in this example!

**Comment by leonardoalt** [comment]
> How hard would it be to test the command lines in the docs? Could we just include a script file that runs all commands (with comments in between instead of markdown), and also test it in the CI?

Yea we should do this at some point

File: book/src/publics.md (hunk 2 above)
Comment (leonardoalt) [reply book/src/publics.md:29]: ah true, just swapped them

**Review by georgwiese — APPROVED** [review]
LGTM!

**Pull request merged**